File: leetcode-350_intersection-of-two-arrays-II.js
//求两个数组的交集[结果元素可重复]
var intersect = function(nums1, nums2) {
    //统计nums1中各个元素出现的次数
    let map = new Map();
    nums1.map(function(n) {
        if (map[n] !== undefined) {
            map[n] += 1;
        } else {
            map[n] = 1;
        }
    });

    //从nums2中取相应次数的元素即是交集
    return nums2.filter(function(n) {
        if (map[n]) {
            map[n]--;
            return true;
        } else {
            return false;
        }
    });
};

intersect([1, 2, 2, 1], [2, 2]);

File: leetcode-350_intersection-of-two-arrays-ii.go
/**
 * 给定两个int数组求交集(元素可重复)
 */
package main

import "fmt"

func main() {
    fmt.Println(intersect([]int{1,2,2,1}, []int{2,2}))
}

func intersect(nums1 []int, nums2 []int) []int {
    m := map[int]int{}
    for _, v := range nums1 { //记录数组一中的元素出现次数
        m[v]++
    }

    ret := []int{}
    for _, v := range nums2 { //从数组二中取回相应次数的相应元素
        if m[v] > 0 {
            ret = append(ret, v)
            m[v]--
        }
    }
    return ret
}
